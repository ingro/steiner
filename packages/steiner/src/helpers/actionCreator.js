// import snakeCase from 'lodash/snakeCase';
import upperFirst from 'lodash/upperFirst';
import defaults from 'lodash/defaults';
import trim from 'lodash/trim';

function addAsyncGroup(resource, actionTypes, group, config) {
    // const upperResource = snakeCase(resource).toUpperCase();
    const upperGroup    = group.toUpperCase();

    resource = trim(resource);

    // const start        = `${upperResource}_${upperGroup}`;
    // const success      = `${upperResource}_${upperGroup}_SUCCESS`;
    // const fail         = `${upperResource}_${upperGroup}_FAIL`;
    const start        = `${resource}/${upperGroup}`;
    const success      = `${resource}/${upperGroup}_SUCCESS`;
    const fail         = `${resource}/${upperGroup}_FAIL`;
    const startAlias   = group;
    const successAlias = `${group}Success`;
    const failAlias    = `${group}Fail`;

    actionTypes[start]   = start;
    actionTypes[success] = success;
    actionTypes[fail]    = fail;

    if (config.addAlias) {
        actionTypes[startAlias]   = start;
        actionTypes[successAlias] = success;
        actionTypes[failAlias]    = fail;
    }
}

export function createActionTypes(resource, config) {
    if (resource == null) {
        throw new Error('Expected resource');
    }

    config = config || {};

    if (config.addAlias == null) {
        config.addAlias = true;
    }

    resource = trim(resource);

    if (resource === '') {
        throw new Error('Expected resource');
    }

    const actionTypes = {};

    addAsyncGroup(resource, actionTypes, 'list', config);
    addAsyncGroup(resource, actionTypes, 'fetch', config);
    addAsyncGroup(resource, actionTypes, 'create', config);
    addAsyncGroup(resource, actionTypes, 'update', config);
    addAsyncGroup(resource, actionTypes, 'delete', config);

    // const upperResource = snakeCase(resource).toUpperCase();

    const resetCurrent = `${resource}/RESET_CURRENT`;
    const changePage   = `${resource}/CHANGE_PAGE`;
    const updateFilter = `${resource}/UDATE_FILTER`;
    const changeOrder  = `${resource}/CHANGE_ORDER`;
    // const inputQuery   = `${upperResource}_INPUT_QUERY`;

    actionTypes.resetCurrent = resetCurrent;
    actionTypes[resetCurrent] = resetCurrent;

    actionTypes.changePage = changePage;
    actionTypes[changePage] = changePage;

    actionTypes.updateFilter = updateFilter;
    actionTypes[updateFilter] = updateFilter;

    actionTypes.changeOrder = changeOrder;
    actionTypes[changeOrder] = changeOrder;

    // actionTypes.inputQuery = inputQuery;
    // actionTypes[inputQuery] = inputQuery;

    return actionTypes;
}

function generateDefaultMessages(resource) {
    return {
        createSuccess: `${resource} created with success!`,
        createFail: `An error occured while creating ${resource}`,
        updateSuccess: `${resource} updated successfully!`,
        updateFail: `An error occured while updating ${resource}`,
        deleteSuccess: `${resource} deleted with success!`,
        deleteFail: `An error occured while deleting ${resource}`
    };
};

function generateSuccessNotification(message, title = 'Hooray!') {
    return {
        title,
        message,
        status: 'success'
    }
}

function generateFailNotification(message, title = 'Oh snap!') {
    return {
        title,
        message,
        status: 'error'
    }
}

export function createActions(resource, actionTypes, messages = {}) {
    if (resource == null) {
        throw new Error('Expected resource');
    }

    resource = trim(upperFirst(resource));

    const defaultMessages = generateDefaultMessages(resource);

    defaults(messages, defaultMessages);

    const actions = {};

    actions[`fetch`] = function(id) {
        return {
            type: actionTypes.fetch,
            payload: {
                id
            }
        };
    }

    actions[`list`] = function(filters) {
        return {
            type: actionTypes.list,
            loadingBar: 'show'
        };
    }

    actions[`create`] = function(data) {
        return {
            type: actionTypes.create,
            payload: {
                data
            }
        }
    }

    actions['createSuccess'] = function(response) {
        return {
            type: actionTypes.createSuccess,
            payload: response,
            notification: generateSuccessNotification(messages.createSuccess)
        }
    }

    actions['createFail'] = function(error) {
        return {
            type: actionTypes.createFail,
            error,
            notification: generateFailNotification(messages.createFail)
        }
    }

    actions[`update`] = function(id, data/*, resolve, reject*/) {
        return {
            type: actionTypes.update,
            payload: {
                id,
                data
                /*resolve,
                reject*/
            }
        }
    }

    actions['updateSuccess'] = function(response) {
        return {
            type: actionTypes.updateSuccess,
            payload: response,
            notification: generateSuccessNotification(messages.updateSuccess)
        }
    }

    actions['updateFail'] = function(error) {
        return {
            type: actionTypes.updateFail,
            error,
            notification: generateFailNotification(messages.updateFail)
        }
    }

    actions[`delete`] = function(id) {
        return {
            type: actionTypes.delete,
            payload: {
                id
            }
        }
    }

    actions['deleteSuccess'] = function(payload) {
        return {
            type: actionTypes.deleteSuccess,
            payload,
            notification: generateSuccessNotification(messages.deleteSuccess)
        }
    }

    actions['deleteFail'] = function(error) {
        return {
            type: actionTypes.deleteFail,
            error,
            notification: generateFailNotification(messages.deleteFail)
        }
    }

    actions[`resetCurrent`] = function() {
        return {
            type: actionTypes.resetCurrent
        };
    }

    // actions[`updateQuery`] = function(q) {
    //     return {
    //         type: actionTypes.inputQuery,
    //         payload: {
    //             q
    //         }
    //     };
    // }

    actions['changePage'] = function(page) {
        return {
            type: actionTypes.changePage,
            payload: {
                page
            }
        }
    }

    actions['changeOrder'] = function(key, direction) {
        return {
            type: actionTypes.changeOrder,
            payload: {
                key,
                direction
            }
        }
    }

    actions['updateFilter'] = function(key, value) {
        return {
            type: actionTypes.updateFilter,
            payload: {
                key,
                value
            }
        };
    }

    return actions;
}

export default {
    createActions,
    createActionTypes
};