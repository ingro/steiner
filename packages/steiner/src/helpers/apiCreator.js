import _ from 'lodash';

export const DEFAULT_PARAMS_MAP = {
    q: 'q',
    _page: 'page',
    _limit: 'perPage',
    _sort: 'orderKey',
    _order: 'orderDirection'
};

export function buildParams(filters, paramsMap = {}) {
    const params = {};

    _.defaults(paramsMap, DEFAULT_PARAMS_MAP);

    _.forOwn(paramsMap, (value, key) => {
        if (typeof value === 'function') {
            params[key] = value(filters);
        } else if (key !== '') {
            params[key] = _.get(filters, value);
        }
    });

    return params;
}

const defaultOptions = {
    omit: []
};

export default function createApi(endpoint, client, paramsMap = {}, options = {}) {
    _.defaults(options, defaultOptions);
    const routes = {};

    const [uri, qs] = endpoint.split('?');

    routes.fetch = function(id) {
        return client({
            url: `/${uri}/${id}${qs ? '?' + qs : ''}`
        });
    }

    routes.list = function(filters) {
        return client({
            url: `/${uri}${qs ? '?' + qs : ''}`,
            params: buildParams(filters, paramsMap)
        });
    }

    routes.create = function(data) {
        return client({
            url: `/${uri}`,
            method: 'post',
            data
        });
    }

    routes.update = function(id, data) {
        return client({
            url: `/${uri}/${id}`,
            method: 'patch',
            data
        });
    }

    routes.delete = function(id) {
        return client({
            url: `/${uri}/${id}`,
            method: 'delete'
        });
    }

    return _.omit(routes, options.omit);
}