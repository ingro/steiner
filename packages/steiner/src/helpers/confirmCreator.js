import { addNotification as notify } from 'reapop';
import _ from 'lodash';

import defaultMessages from '../messages/en';

const defaultConfirmOptions = {
    status: 'warning',
    position: 'tc'
};

export default function createConfirm(options = {}) {
    _.defaults(options, defaultConfirmOptions, defaultMessages.messages.confirmDialog);

    return notify({
        title: options.title,
        message: options.message,
        status: options.status,
        dismissible: false,
        dismissAfter: 0,
        position: options.position,
        buttons: [
            {
                name: options.btn_primary_text,
                primary: true,
                onClick: () => {
                    options.onSuccess();
                }
            },
            {
                name: options.btn_secondary_text,
                primary: false
            }
        ]
    });
}