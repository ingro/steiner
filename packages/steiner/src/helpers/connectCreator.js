import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

export function connectList(component, actions, selectors) {
    function mapStateToProps(state) {
        const list = selectors.listSelector(state);

        return {
            ...list,
            items: selectors.itemsSelector(state),
            selected: selectors.getSelectedId(state),
            component
        };
    }

    function mapDispatchToProps(dispatch) {
        const boundedActions = bindActionCreators(actions, dispatch);

        return {
            ...boundedActions,
            dispatch
        }
    }

    return connect(mapStateToProps, mapDispatchToProps)(component);
}

export function connectItem(component, actions, selectors) {
    function mapStateToProps(state, ownProps) {
        const current = selectors.currentSelector(state);
        const isNew = ownProps.params.id === 'create';

        return {
            ...current,
            item: isNew ? {} : current.item
        };
    }

    function mapDispatchToProps(dispatch) {
        return bindActionCreators(actions, dispatch);
    }

    return connect(mapStateToProps, mapDispatchToProps)(component);
}

export default {
    connectItem,
    connectList
};