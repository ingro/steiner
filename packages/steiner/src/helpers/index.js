import actionCreatorModule from './actionCreator';
import apiCreatorModule from './apiCreator';
import connectCreatorModule from './connectCreator';
import formHelperModule from './formHelper';
import notificationMiddlewareModule from './notificationMiddleware';
import reducerCreatorModule from './reducerCreator';
import reduxFormSagaModule from './reduxFormSaga';
import routeCreatorModule from './routeCreator';
import sagaCreatorModule from './sagaCreator';

export const actionCreator = actionCreatorModule;
export const apiCreator = apiCreatorModule;
export const connectCreator = connectCreatorModule;
export const formHelper = formHelperModule;
export const notificationMiddleware = notificationMiddlewareModule;
export const reducerCreator = reducerCreatorModule;
export const reduxFormSaga = reduxFormSagaModule;
export const routeCreator = routeCreatorModule;
export const sagaCreator = sagaCreatorModule;