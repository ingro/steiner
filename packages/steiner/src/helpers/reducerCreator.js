import Immutable from 'seamless-immutable';
import matchSorter from 'match-sorter';
import sift from 'sift';
import _ from 'lodash';

import { actionTypes as authActionTypes } from '../auth/actions';

export const DEFAULT_STATE = Immutable({
    list: {
        errorMessage: null,
        isFetching: false,
        itemsById: {},
        itemsId: [],
        selected: [],
        filters: {
            q: '',
            page: 1,
            perPage: 20,
            orderKey: null,
            orderDirection: 'ASC'
        },
        total: 0
    },
    current: {
        errorMessage: null,
        isFetching: false,
        item: null
    }
});

const defaultListSuccessOptions = {
    items: 'data.data',
    total: 'data.meta.total',
    idKey: 'id'
};

function createListSuccessHandler(options = {}) {
    _.defaults(options, defaultListSuccessOptions);

    return function listSuccess(state, action) {
        const items = typeof options.items === 'function' ? options.items(action.payload) : _.get(action.payload, options.items);

        const itemsId = items.map(item => item[options.idKey]);

        return state.update('list', list => ({
            ...list,
            itemsId,
            itemsById: items.reduce((byId, item) => {
                byId[item[options.idKey]] = item;

                return byId;
            }, {}),
            selected: _.intersection(itemsId, state.list.selected),
            isFetching: false,
            errorMessage: null,
            total: typeof options.total === 'function' ? options.total(action.payload) : _.get(action.payload, options.total)
        }));
    }
}

function list(state, action) {
    return state.setIn(['list', 'isFetching'], true);
}

function listFail(state, action) {
    return state.update('list', list => ({
        ...list,
        itemsById: {},
        itemsId: [],
        total: 0,
        isFetching: false,
        errorMessage: action.error.message
    }));
}

function fetch(state, action) {
    return state.update('current', current => ({
        ...current,
        item: null,
        isFetching: true
    }));
}

function fetchSuccess(state, action) {
    return state.update('current', () => ({
        item: action.payload.data,
        isFetching: false,
        errorMessage: null
    }));
}

function fetchFail(state, action) {
    return state.update('current', current => ({
        ...current,
        isFetching: false,
        errorMessage: action.error.message
    }));
}

function deleteSuccess(state, action) {
    return state.updateIn(['list'], list => ({
        ...list,
        itemsById: list.itemsById.without(action.payload.id),
        itemsId: _.difference(list.itemsId, [action.payload.id])
    }));
}

function resetCurrent(state, action) {
    return state.update('current', () => ({
        item: null,
        isFetching: false,
        errorMessage: null
    }));
}

function updateFilter(state, action) {
    return state.setIn(['list', 'filters', action.payload.key], action.payload.value);
}

function setFilters(state, action) {
    const toSync = action.payload;

    if (toSync.page) {
        toSync.page = parseInt(toSync.page, 10);
    }

    if (toSync.perPage) {
        toSync.perPage = parseInt(toSync.perPage, 10);
    }

    return state.updateIn(['list', 'filters'], filters => {
        return Immutable.merge(filters, toSync);
        // ...filters,
        // ...toSync
    }); 
}

// function resetFilters(state, action, defaultState) {
//     return state.setIn(['list', 'filters'], defaultState.list.filters);
// }

function changePage(state, action) {
    return state.setIn(['list', 'filters', 'page'], parseInt(action.payload.page, 10));
}

function changeOrder(state, action) {
    const { key, direction } = action.payload;

    return state.updateIn(['list', 'filters'], filters => {
        return Immutable.merge(filters, {
            orderKey: key,
            orderDirection: direction
        });
        // ...filters,
        // orderKey: key,
        // orderDirection: direction
    });
}

function select(state, action) {
    const selected = _.uniq([].concat(state.list.selected, action.payload));
    return state.setIn(['list', 'selected'], selected);
}

function deselect(state, action) {
    const selected = _.difference(state.list.selected, action.payload);
    return state.setIn(['list', 'selected'], selected);
}

function selectAll(state) {
    return state.setIn(['list', 'selected'], state.list.itemsId);
}

function deselectAll(state) {
    return state.setIn(['list', 'selected'], []);
}

/**
 * For example
    {
        additionalFilters: {
            countryId: {
                field: 'countryId',
                op: '$eq'
            },
            priceMin: {
                field: 'price',
                op: '$gte',
            },
            priceMax: {
                field: 'price',
                op: '$lte',
            }
        },
        searchKeys: ['title']
    }
 */
const defaultFilterCollectionOptions = {
    additionalFilters: {},
    defaultOrderKey: 'id',
    searchKeys: ['name']
};

function createFilterCollectionHandler(options = {}) {
    _.defaults(options, defaultFilterCollectionOptions);

    const { additionalFilters, defaultOrderKey, searchKeys } = options;

    return function filterCollection(state, action) {
        // let itemsId = [];
        let filtered = [];

        const filters = state.list.filters;
        const items = state.list.itemsById;

        if (filters.q === '') {
            filtered = Object.values(items);
            // itemsId = Object.keys(items);
        } else {
            filtered = matchSorter(Object.values(items), filters.q, {
                keys: searchKeys,
                threshold: matchSorter.rankings.CONTAINS
            });

            // itemsId = matches.map(match => match.id);
        }

        // let filtered = itemsId.map(id => items[id]);

        const query = {};

        Object.keys(filters).forEach(key => {
            if (_.includes(Object.keys(additionalFilters), key) && filters[key]) {
                const queryKey = additionalFilters[key].field;
                if (! query[queryKey]) {
                    query[queryKey] = {};
                }

                query[queryKey][additionalFilters[key].op] = filters[key];
            }
        });
        
        filtered = sift(query, filtered);

        let sorted = _.sortBy(filtered, filters.orderKey ? filters.orderKey : defaultOrderKey);

        if (filters.orderDirection === 'DESC') {
            sorted = sorted.reverse();
        }

        if (filters.perPage > 0 && sorted.length > 0) {
            const start = (filters.page * filters.perPage) - filters.perPage;
            const end = filters.page * filters.perPage;
            sorted = sorted.slice(start, end);
        }
        
        return state.updateIn(['list'], list => ({
            ...list,
            itemsId: sorted.map(item => item.id)
        }));
    }
}

export function createHandlers(actionTypes, options = {}) {
    return {
        [actionTypes.list]: list,
        [actionTypes.listSuccess]: createListSuccessHandler(options),
        [actionTypes.listFail]: listFail,
        [actionTypes.fetch]: fetch,
        [actionTypes.fetchSuccess]: fetchSuccess,
        [actionTypes.fetchFail]: fetchFail,
        [actionTypes.deleteSuccess]: deleteSuccess,
        [actionTypes.resetCurrent]: resetCurrent,
        [actionTypes.updateFilter]: updateFilter,
        [actionTypes.setFilters]: setFilters,
        // [actionTypes.resetFilters]: resetFilters,
        [actionTypes.changePage]: changePage,
        [actionTypes.changeOrder]: changeOrder,
        [actionTypes.select]: select,
        [actionTypes.deselect]: deselect,
        [actionTypes.selectAll]: selectAll,
        [actionTypes.deselectAll]: deselectAll,
        [actionTypes.filterCollection]: createFilterCollectionHandler(options)
    };
}

export function createSelectors(key) {
    return {
        currentSelector: state => state[key].current,
        getFilters: state => state[key].list.filters,
        getSelected: state => state[key].list.selected.map(id => state[key].list.itemsById[id]),
        getSelectedId: state => state[key].list.selected,
        itemSelector: (state, id) => state[key].list.itemsById[id],
        itemsSelector: state => state[key].list.itemsId.map(id => state[key].list.itemsById[id]),
        listSelector: state => state[key].list
    };
}

export function createReducer(handlers, defaultState = DEFAULT_STATE, options = {}) {
    _.defaults(options, {
        resetOnLogout: true
    });

    if (options.resetOnLogout) {
        handlers[authActionTypes.logoutSuccess] = () => defaultState;
    }

    return function(state = defaultState, action) {
        if (handlers.hasOwnProperty(action.type)) {
            return handlers[action.type](state, action/*, DEFAULT_STATE*/);
        } else {
            return state;
        }
    }
}

export default {
    createHandlers,
    createReducer,
    createSelectors,
    DEFAULT_STATE
};