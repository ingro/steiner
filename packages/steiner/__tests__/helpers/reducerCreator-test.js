import { createHandlers, createSelectors, DEFAULT_STATE } from '../../src/helpers/reducerCreator';
import { createActionTypes } from '../../src/helpers/actionCreator';

describe('createHandlers', () => {
    const actionTypes = createActionTypes('posts');
    const handlers = createHandlers(actionTypes);

    const state = DEFAULT_STATE;

    it('handles correctly list action', () => {
        expect(handlers[actionTypes.list](state)).toMatchSnapshot();
    });

    it('handles correctly listSuccess action', () => {
        const action = {
            payload: {
                data: {
                    data: [
                        {
                            id: 33,
                            name: 'foo'
                        },
                        {
                            id: 44,
                            name: 'bar'
                        }
                    ],
                    meta: {
                        total: 2
                    }
                }
            }
        }

        expect(handlers[actionTypes.listSuccess](state, action)).toMatchSnapshot();
    });

    it('handles correctly listFail action', () => {
        const action = {
            error: {
                message: 'error'
            }
        };

        expect(handlers[actionTypes.listFail](state, action)).toMatchSnapshot();
    });

    it('handles correctly fetch action', () => {
        expect(handlers[actionTypes.fetch](state)).toMatchSnapshot();
    });

    it('handles correctly fetchSuccess action', () => {
        const action = {
            payload: {
                data: {
                    id: 42,
                    name: 'foo'
                }
            }
        };

        expect(handlers[actionTypes.fetchSuccess](state, action)).toMatchSnapshot();
    });

    it('handles correctly fetchFail action', () => {
        const action = {
            error: {
                message: 'error'
            }
        };

        expect(handlers[actionTypes.fetchFail](state, action)).toMatchSnapshot();
    });

    it('handles correctly deleteSuccess action', () => {
        const action = {
            payload: {
                id: 42
            }
        };

        const deleteState = DEFAULT_STATE.setIn(['list', 'itemsById'], { 33: { foo: 'bar' }, 42: { foo: 'bar' }}).setIn(['list', 'itemsId'], [33, 42]);

        expect(handlers[actionTypes.deleteSuccess](deleteState, action)).toMatchSnapshot();
    });

    it('handles correctly resetCurrent action', () => {
        const resetState = DEFAULT_STATE.setIn(['current', 'item'], { id: 42, foo: 'bar'});

        expect(handlers[actionTypes.resetCurrent](resetState)).toMatchSnapshot();
    });

    it('handles correctly updateFilter action', () => {
        const action = {
            payload: {
                key: 'q',
                value: 'hello'
            }
        };

        expect(handlers[actionTypes.updateFilter](state, action)).toMatchSnapshot();
    });

    it('handles correctly setFilters action', () => {
        const action = {
            payload: {
                page: '10',
                q: 'hello'
            }
        };

        expect(handlers[actionTypes.setFilters](state, action)).toMatchSnapshot();
    });

    it('handles correctly changePage action', () => {
        const action = {
            payload: {
                page: '10'
            }
        };

        expect(handlers[actionTypes.changePage](state, action)).toMatchSnapshot();
    });

    it('handles correctly changeOrder action', () => {
        const action = {
            payload: {
                key: 'name',
                direction: 'DESC'
            }
        };

        expect(handlers[actionTypes.changeOrder](state, action)).toMatchSnapshot();
    });

    it('handles correctly select action', () => {
        const action = {
            payload: [33]
        };

        const selectState = DEFAULT_STATE.setIn(['list', 'selected'], [42]);

        expect(handlers[actionTypes.select](selectState, action)).toMatchSnapshot();
    });

    it('handles correctly deselect action', () => {
        const action = {
            payload: [33]
        };

        const deselectState = DEFAULT_STATE.setIn(['list', 'selected'], [33]);

        expect(handlers[actionTypes.deselect](deselectState, action)).toMatchSnapshot();
    });

    it('handles correctly selectAll action', () => {
        const selectState = DEFAULT_STATE.setIn(['list', 'itemsId'], [33, 42, 56]);

        expect(handlers[actionTypes.selectAll](selectState)).toMatchSnapshot();
    });

    it('handles correctly deselectAll action', () => {
        const selectState = DEFAULT_STATE.setIn(['list', 'selected'], [33, 42, 56]);

        expect(handlers[actionTypes.deselectAll](selectState)).toMatchSnapshot();
    });

    // it('handles correctly resetFilters action', () => {
    //     const prevState = DEFAULT_STATE.setIn(['list', 'filters'], {
    //         q: 'foo',
    //         page: 2,
    //         perPage: 20,
    //         orderKey: null,
    //         orderDirection: 'ASC'
    //     });

    //     expect(handlers[actionTypes.resetFilters](prevState, {}, DEFAULT_STATE)).toMatchSnapshot();
    // });
});

describe('createSelectors', () => {
    const selectors = createSelectors('posts');

    const state = {
        posts: DEFAULT_STATE
            .setIn(['list', 'itemsId'], [33, 42])
            .setIn(['list', 'itemsById'], { 33: { id: 33, name: 'foo' }, 42: { id: 42, name: 'bar' }})
            .setIn(['list', 'selected'], [42])
    }

    it('listSelector returns the list from the state', () => {
        expect(selectors.listSelector(state)).toMatchSnapshot();
    });

    it('itemsSelector returns the list of items from the state', () => {
        expect(selectors.itemsSelector(state)).toMatchSnapshot();
    });

    it('currentSelector returns the current item from the state', () => {
        expect(selectors.currentSelector(state)).toMatchSnapshot();
    });

    it('getFilters returns the filters from the state', () => {
        expect(selectors.getFilters(state)).toMatchSnapshot();
    });

    it('getSelectedId returns the filters from the state', () => {
        expect(selectors.getSelectedId(state)).toMatchSnapshot();
    });

    it('getSelected returns the filters from the state', () => {
        expect(selectors.getSelected(state)).toMatchSnapshot();
    });
});