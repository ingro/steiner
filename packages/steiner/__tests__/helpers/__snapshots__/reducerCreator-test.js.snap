// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createHandlers handles correctly changeOrder action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "DESC",
      "orderKey": "name",
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly changePage action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 10,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly deleteSuccess action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {
      "33": Object {
        "foo": "bar",
      },
    },
    "itemsId": Array [
      33,
    ],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly deselect action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly deselectAll action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly fetch action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": true,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly fetchFail action 1`] = `
Object {
  "current": Object {
    "errorMessage": "error",
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly fetchSuccess action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": Object {
      "id": 42,
      "name": "foo",
    },
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly list action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": true,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly listFail action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": "error",
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly listSuccess action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {
      "33": Object {
        "id": 33,
        "name": "foo",
      },
      "44": Object {
        "id": 44,
        "name": "bar",
      },
    },
    "itemsId": Array [
      33,
      44,
    ],
    "selected": Array [],
    "total": 2,
  },
}
`;

exports[`createHandlers handles correctly resetCurrent action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly select action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [
      42,
      33,
    ],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly selectAll action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [
      33,
      42,
      56,
    ],
    "selected": Array [
      33,
      42,
      56,
    ],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly setFilters action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 10,
      "perPage": 20,
      "q": "hello",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createHandlers handles correctly updateFilter action 1`] = `
Object {
  "current": Object {
    "errorMessage": null,
    "isFetching": false,
    "item": null,
  },
  "list": Object {
    "errorMessage": null,
    "filters": Object {
      "orderDirection": "ASC",
      "orderKey": null,
      "page": 1,
      "perPage": 20,
      "q": "hello",
    },
    "isFetching": false,
    "itemsById": Object {},
    "itemsId": Array [],
    "selected": Array [],
    "total": 0,
  },
}
`;

exports[`createSelectors currentSelector returns the current item from the state 1`] = `
Object {
  "errorMessage": null,
  "isFetching": false,
  "item": null,
}
`;

exports[`createSelectors getFilters returns the filters from the state 1`] = `
Object {
  "orderDirection": "ASC",
  "orderKey": null,
  "page": 1,
  "perPage": 20,
  "q": "",
}
`;

exports[`createSelectors getSelected returns the filters from the state 1`] = `
Array [
  Object {
    "id": 42,
    "name": "bar",
  },
]
`;

exports[`createSelectors getSelectedId returns the filters from the state 1`] = `
Array [
  42,
]
`;

exports[`createSelectors itemsSelector returns the list of items from the state 1`] = `
Array [
  Object {
    "id": 33,
    "name": "foo",
  },
  Object {
    "id": 42,
    "name": "bar",
  },
]
`;

exports[`createSelectors listSelector returns the list from the state 1`] = `
Object {
  "errorMessage": null,
  "filters": Object {
    "orderDirection": "ASC",
    "orderKey": null,
    "page": 1,
    "perPage": 20,
    "q": "",
  },
  "isFetching": false,
  "itemsById": Object {
    "33": Object {
      "id": 33,
      "name": "foo",
    },
    "42": Object {
      "id": 42,
      "name": "bar",
    },
  },
  "itemsId": Array [
    33,
    42,
  ],
  "selected": Array [
    42,
  ],
  "total": 0,
}
`;
